### PR TITLE
Added Telemetry markers to Chat SDK 

### PR DESCRIPTION
**Pull request by anthonyyemitan**
Added Telemetry markers to the following functions: 

- SendTypingEvent
- OnAgentEndSession
- UploadFileAttachment
- DownloadFileAttachment
- EmailLiveChatTranscript
- GetLiveChatTranscript
- GetVoiceVideoCalling ,
- onCallDisconnected = "onCallDisconnected"

### CODE REVIEW

File: src/OmnichannelChatSDK.ts [hunk 1]
```diff
@@ -484,46 +510,91 @@ class OmnichannelChatSDK {
...
+                emailTranscriptOptionalParams.authenticatedUserToken = this.authenticatedUserToken;
+            }
+            const emailRequestBody = {
+                ChatId: this.chatToken.chatId,
+                EmailAddress: body.emailAddress,
+                DefaultAttachmentMessage: body.attachmentMessage,
+                CustomerLocale: body.locale
+            };
+
+            this.scenarioMarker.completeScenario(TelemetryEvent.EmailLiveChatTranscript);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:559]: I'd not complete the scenario before `OCClient.emailTranscript()` has even completed. 

I'd do:

```
const response = this.OCClient.emailTranscript(...);
this.scenarioMarker.completeScenario(TelemetryEvent.EmailLiveChatTranscript);
return response;
```

File: src/OmnichannelChatSDK.ts [hunk 2]
```diff
@@ -484,46 +510,91 @@ class OmnichannelChatSDK {
...
+            RequestId: this.requestId,
+            chatId: this.chatToken.chatId as string
+        });
+
+        try {
+            if (this.authenticatedUserToken) {
+                getChatTranscriptOptionalParams.authenticatedUserToken = this.authenticatedUserToken;
+            }
+
+            this.scenarioMarker.completeScenario(TelemetryEvent.EmailLiveChatTranscript);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:586]: Same issue as with `emailLiveChatTranscript()`

File: src/OmnichannelChatSDK.ts [hunk 3]
```diff
@@ -639,7 +713,10 @@ class OmnichannelChatSDK {
     private async getIC3Client() {
         if (platform.isNode() || platform.isReactNative()) {
             this.debug && console.debug('IC3Core');
-            this.scenarioMarker.startScenario(TelemetryEvent.GetIC3Client);
+            this.scenarioMarker.startScenario(TelemetryEvent.GetIC3Client, {
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:716]: `GetIC3Client` is already done. See [here](https://github.com/microsoft/omnichannel-chat-sdk/blob/f9059fb6eec2623470efb30fce52e630a4a7d9dc/src/OmnichannelChatSDK.ts#L679)

It's not needed when platform is `Node` or `React Native`.

File: src/OmnichannelChatSDK.ts (hunk 2 above)
Comment (xTEddie) [reply src/OmnichannelChatSDK.ts:586]: Scenario should be `TelemetryEvent.GetLiveChatTranscript` and not `TelemetryEvent.EmailLiveChatTranscript`

File: src/OmnichannelChatSDK.ts [hunk 4]
```diff
@@ -484,46 +510,91 @@ class OmnichannelChatSDK {
...
     }
 
     public async getLiveChatTranscript(): Promise<any> { // eslint-disable-line @typescript-eslint/no-explicit-any
         const getChatTranscriptOptionalParams: IGetChatTranscriptsOptionalParams = {};
-        if (this.authenticatedUserToken) {
-            getChatTranscriptOptionalParams.authenticatedUserToken = this.authenticatedUserToken;
+
+        this.scenarioMarker.startScenario(TelemetryEvent.GetLiveChatTranscript, {
+            RequestId: this.requestId,
+            chatId: this.chatToken.chatId as string
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:578]: `chatId` should be `ChatId` or a new column will be created in the logs

File: src/OmnichannelChatSDK.ts [hunk 5]
```diff
@@ -589,7 +660,10 @@ class OmnichannelChatSDK {
                     SpoolSDK: spoolSDKCDNUrl
                 });
 
-                this.scenarioMarker.startScenario(TelemetryEvent.GetVoiceVideoCalling);
+                this.scenarioMarker.startScenario(TelemetryEvent.GetVoiceVideoCalling, {            
+                    RequestId: this.requestId,
+                    chatId: this.chatToken.chatId as string
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:665]: `chatId` should be `ChatId`

File: src/OmnichannelChatSDK.ts [hunk 6]
```diff
@@ -639,7 +713,10 @@ class OmnichannelChatSDK {
     private async getIC3Client() {
         if (platform.isNode() || platform.isReactNative()) {
             this.debug && console.debug('IC3Core');
-            this.scenarioMarker.startScenario(TelemetryEvent.GetIC3Client);
+            this.scenarioMarker.startScenario(TelemetryEvent.GetIC3Client, {
+                RequestId: this.requestId,
+                chatId: this.chatToken.chatId as string
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:718]: `chatId` should be `ChatId`

File: src/api/createVoiceVideoCalling.ts [hunk 7]
```diff
@@ -340,11 +340,15 @@ export class VoiceVideoCallingProxy {
                 SecondaryChannelEventType: SecondaryChannelEvents.End
             }
 
+            this.scenarioMarker?.startScenario(TelemetryEvent.onCallDisconnected);
```
Comment (xTEddie) [inline src/api/createVoiceVideoCalling.ts:343]: We need to pass `callId`. You can refer to [this](https://github.com/microsoft/omnichannel-chat-sdk/blob/f9059fb6eec2623470efb30fce52e630a4a7d9dc/src/api/createVoiceVideoCalling.ts#L135)

```
        this.scenarioMarker?.startScenario(params.withVideo? TelemetryEvent.AcceptVideoCall: TelemetryEvent.AcceptVoiceCall, {
            CallId: callId || ''
        });
```

File: src/api/createVoiceVideoCalling.ts [hunk 8]
```diff
@@ -340,11 +340,15 @@ export class VoiceVideoCallingProxy {
...
+
             try {
                 this.callingParams?.OCClient.makeSecondaryChannelEventRequest(this.callingParams?.chatToken.requestId, body);
                 this.debug && console.debug(`[VoiceVideoCallingProxy][onCallDisconnected][makeSecondaryChannelEventRequest]`);
+                this.scenarioMarker?.completeScenario(TelemetryEvent.onCallDisconnected);
```
Comment (xTEddie) [inline src/api/createVoiceVideoCalling.ts:348]: Need to pass `callId`

File: src/api/createVoiceVideoCalling.ts [hunk 9]
```diff
@@ -340,11 +340,15 @@ export class VoiceVideoCallingProxy {
...
             } catch (e) {
                 console.error(`[VoiceVideoCallingProxy][onCallDisconnected][makeSecondaryChannelEventRequest] Failure ${e}`);
+                this.scenarioMarker?.failScenario(TelemetryEvent.onCallDisconnected);
```
Comment (xTEddie) [inline src/api/createVoiceVideoCalling.ts:351]: We need to pass `callId` & `ExceptionDetails`. You can refer to [this](https://github.com/microsoft/omnichannel-chat-sdk/blob/f9059fb6eec2623470efb30fce52e630a4a7d9dc/src/api/createVoiceVideoCalling.ts#L170)

File: src/OmnichannelChatSDK.ts (hunk 3 above)
Comment (anthonyyemitan) [reply src/OmnichannelChatSDK.ts:760]: Do we not need the RequestID or ChatID?

Comment (xTEddie) [reply src/OmnichannelChatSDK.ts:760]: No, `GetIC3Client` is used to the load the IC3 library. At this point in time, no chat has even started

File: src/api/createVoiceVideoCalling.ts [hunk 10]
```diff
@@ -340,11 +340,26 @@ export class VoiceVideoCallingProxy {
...
             }
 
+            this.scenarioMarker?.startScenario(TelemetryEvent.OnCallDisconnected, {
+                CallId: this.callId || ''
+            });
+
             try {
                 this.callingParams?.OCClient.makeSecondaryChannelEventRequest(this.callingParams?.chatToken.requestId, body);
                 this.debug && console.debug(`[VoiceVideoCallingProxy][onCallDisconnected][makeSecondaryChannelEventRequest]`);
+                this.scenarioMarker?.completeScenario(TelemetryEvent.OnCallDisconnected,{
```
Comment (xTEddie) [inline src/api/createVoiceVideoCalling.ts:350]: nit: space after `,`

File: src/OmnichannelChatSDK.ts [hunk 11]
```diff
@@ -465,13 +464,24 @@ class OmnichannelChatSDK {
...
+            ChatId: this.chatToken.chatId as string
+        });
+
         try {
             await this.conversation!.indicateTypingStatus(0);
             const members: IPerson[] = await this.conversation!.getMembers();
             const botMembers = members.filter((member: IPerson) => member.type === PersonType.Bot);
             await this.conversation!.sendMessageToBot(botMembers[0].id, {payload: typingPayload});
+
+            this.scenarioMarker.completeScenario(TelemetryEvent.SendTypingEvent);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:479]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 12]
```diff
@@ -465,13 +464,24 @@ class OmnichannelChatSDK {
...
+
         } catch (error) {
             console.error("OmnichannelChatSDK/sendTypingEvent/error");
+
+            this.scenarioMarker.failScenario(TelemetryEvent.SendTypingEvent);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:484]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 13]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
 
         try {
             await this.conversation!.sendFileMessage(fileMetadata, messageToSend);
+            this.scenarioMarker.completeScenario(TelemetryEvent.UploadFileAttachment);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:557]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 14]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
             return messageToSend;
         } catch (error) {
             console.error(`OmnichannelChatSDK/uploadFileAttachment/error: ${error}`);
+            this.scenarioMarker.failScenario(TelemetryEvent.UploadFileAttachment);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:561]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 15]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
     public async downloadFileAttachment(fileMetadata: IFileMetadata): Promise<Blob> {
-        return this.conversation!.downloadFile(fileMetadata);
+        this.scenarioMarker.startScenario(TelemetryEvent.DownloadFileAttachment, {
+            RequestId: this.requestId,
+            ChatId: this.chatToken.chatId as string
+        })
+
+        try { 
+            const downloadedFile = this.conversation!.downloadFile(fileMetadata);
+            this.scenarioMarker.completeScenario(TelemetryEvent.DownloadFileAttachment);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:574]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 16]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
+            return downloadedFile;
+        }
+        catch(e) {
+            this.scenarioMarker.failScenario(TelemetryEvent.DownloadFileAttachment);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:578]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 17]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
 
     public async downloadFileAttachment(fileMetadata: IFileMetadata): Promise<Blob> {
-        return this.conversation!.downloadFile(fileMetadata);
+        this.scenarioMarker.startScenario(TelemetryEvent.DownloadFileAttachment, {
+            RequestId: this.requestId,
+            ChatId: this.chatToken.chatId as string
+        })
+
+        try { 
+            const downloadedFile = this.conversation!.downloadFile(fileMetadata);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:573]: Missing `await`

File: src/OmnichannelChatSDK.ts [hunk 18]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
+                getChatTranscriptOptionalParams.authenticatedUserToken = this.authenticatedUserToken;
+            }
+            
+            const transcriptResponse = this.OCClient.getChatTranscripts(
+                this.requestId,
+                this.chatToken.chatId,
+                this.chatToken.token,
+                getChatTranscriptOptionalParams);
+
+            this.scenarioMarker.completeScenario(TelemetryEvent.GetLiveChatTranscript);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:638]: Missing ChatId & RequestId

File: src/OmnichannelChatSDK.ts [hunk 19]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
+            return transcriptResponse;      
         }
-        return this.OCClient.getChatTranscripts(
-            this.requestId,
-            this.chatToken.chatId,
-            this.chatToken.token,
-            getChatTranscriptOptionalParams);
-    }
+        catch(e) {
+            this.scenarioMarker.failScenario(TelemetryEvent.GetLiveChatTranscript);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:643]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 20]
```diff
@@ -492,17 +502,33 @@ class OmnichannelChatSDK {
...
+        try {
+            this.conversation?.registerOnThreadUpdate((message: IRawThread) => {
+                const {members} = message;
+
+                // Agent ending conversation would have 1 member left in the chat thread
+                if (members.length === 1) {
+                    onAgentEndSessionCallback(message);
+                }
+            });
+            this.scenarioMarker.completeScenario(TelemetryEvent.OnAgentEndSession);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:519]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 21]
```diff
@@ -492,17 +502,33 @@ class OmnichannelChatSDK {
...
+        }
+        catch(e) {
+            this.scenarioMarker.failScenario(TelemetryEvent.OnAgentEndSession);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:522]: Missing `ChatId` & `RequestId`

File: src/OmnichannelChatSDK.ts [hunk 22]
```diff
@@ -528,46 +554,95 @@ class OmnichannelChatSDK {
...
+        })
+
+        try { 
+            const downloadedFile = this.conversation!.downloadFile(fileMetadata);
+            this.scenarioMarker.completeScenario(TelemetryEvent.DownloadFileAttachment);
+            return downloadedFile;
+        }
+        catch(e) {
+            this.scenarioMarker.failScenario(TelemetryEvent.DownloadFileAttachment);
+            return this.conversation!.downloadFile(fileMetadata);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:579]: We should return `new Error('DownloadFileAttachmentFailed')` on failure

File: src/api/createVoiceVideoCalling.ts [hunk 23]
```diff
@@ -340,11 +340,26 @@ export class VoiceVideoCallingProxy {
...
+                    CallId: this.callId || ''
+                });
             } catch (e) {
                 console.error(`[VoiceVideoCallingProxy][onCallDisconnected][makeSecondaryChannelEventRequest] Failure ${e}`);
+                
+                const exceptionDetails = {
+                    response: 'OnCallDisconnectedFailed'
+                }
+                this.scenarioMarker?.failScenario(TelemetryEvent.OnCallDisconnected,{
```
Comment (xTEddie) [inline src/api/createVoiceVideoCalling.ts:359]: nit: space after `,`

File: src/api/createVoiceVideoCalling.ts [hunk 24]
```diff
@@ -340,11 +340,26 @@ export class VoiceVideoCallingProxy {
...
                 this.callingParams?.OCClient.makeSecondaryChannelEventRequest(this.callingParams?.chatToken.requestId, body);
                 this.debug && console.debug(`[VoiceVideoCallingProxy][onCallDisconnected][makeSecondaryChannelEventRequest]`);
+                this.scenarioMarker?.completeScenario(TelemetryEvent.OnCallDisconnected,{
+                    CallId: this.callId || ''
+                });
             } catch (e) {
                 console.error(`[VoiceVideoCallingProxy][onCallDisconnected][makeSecondaryChannelEventRequest] Failure ${e}`);
+                
+                const exceptionDetails = {
+                    response: 'OnCallDisconnectedFailed'
```
Comment (xTEddie) [inline src/api/createVoiceVideoCalling.ts:357]: It's the OCClient.makeSecondaryChannelEventRequest that's failing. So `response` should be `OCClientMakeSecondaryChannelEventRequestFailed`

File: src/OmnichannelChatSDK.ts (hunk 22 above)
Comment (anthonyyemitan) [reply src/OmnichannelChatSDK.ts:686]: Syntax did not work, but I found something online that does the same thing

File: src/OmnichannelChatSDK.ts [hunk 25]
```diff
@@ -712,7 +821,10 @@ class OmnichannelChatSDK {
                     SpoolSDK: spoolSDKCDNUrl
                 });
 
-                this.scenarioMarker.startScenario(TelemetryEvent.GetVoiceVideoCalling);
+                this.scenarioMarker.startScenario(TelemetryEvent.GetVoiceVideoCalling, {
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:824]: Please remove `RequestId` & `ChatId` since they may not be relevant at this point of time.

**Review by xTEddie — APPROVED** [review]

**Pull request merged**